### PR TITLE
Revised Type class to better document BindingFlags behavior

### LINKED ISSUE #2052: Type.GetProperties doc need to fixes
The doc 

https://docs.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=netframework-4.7.2#System_Type_GetProperties_System_Reflection_BindingFlags_

need some fixes to address the suggestions mentioned in the community report  https://developercommunity.visualstudio.com/content/problem/463525/cannot-see-public-property-in-reflection.html?childToView=498831#reply-form

here is the suggestion text from that report 

```
Thank you for your reply. May I suggest that you consider rewriting the help for this?

The sentence in the comment "You must specify either BindingFlags.Instance or BindingFlags.Static in order to get a return." that you quoted uses the term "or", which I instinctively interpreted in its Boolean sense (being an article written for programmers). May I suggest that you add the phrase "but not both" to eliminate the confusion?

Also, I would like to point you to the following block of text which occurs more than once on that page:

A property is considered public to reflection if it has at least one accessor that is public. Otherwise the property is considered private, and you must use BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static (in Visual Basic, combine the values using Or) to get it.  
You may notice that it uses ```Instance``` and Static bindings in the same request, implying that this is a normal thing to do. If it is only legal to use Static and Instance together when also used with NonPublic, then you need to state this explicitly, to avoid confusing this reader.

```

### PR DESCRIPTION
**Pull request by rpetrusha**
## Revised to document BindingFlags behavior

This PR corrects a number of inaccuracies in the Type documentation:

- BindingFlags.Static and/or BindingFlags.Instance are not sufficient to get a return value. Either BindingFlags.Public and/or BindingFlags.NonPublic must also be used.
- The methods that return an array don't return null.
- Supplying a value of 0 for a BindingFlags enumeration member seems confusing. I've replaced it with BindingFlags.Default.
- Numerous other changes.

Fixes #2052 

//cc @tarekgh @GrabYourPitchforks

### CODE REVIEW

**Comment by rpetrusha** [comment]
Closing and reopening to begin new build.

**Comment by tarekgh** [comment]
@steveharter may help reviewing the changes here.

File: xml/System/Type.xml [hunk 1]
```diff
@@ -2426,43 +2431,37 @@ TopNamespace.Sub\+Namespace.ContainingClass+NestedClass, MyAssembly, Version=1.3
         <Parameter Name="bindingAttr" Type="System.Reflection.BindingFlags" Index="0" FrameworkAlternate="netcore-2.0;netcore-2.1;netcore-2.2;netcore-3.0;netframework-1.1;netframework-2.0;netframework-3.0;netframework-3.5;netframework-4.0;netframework-4.5;netframework-4.5.1;netframework-4.5.2;netframework-4.6;netframework-4.6.1;netframework-4.6.2;netframework-4.7;netframework-4.7.1;netframework-4.7.2;netframework-4.8;netstandard-2.0;xamarinandroid-7.1;xamarinios-10.8;xamarinmac-3.0;netstandard-2.1" />
       </Parameters>
       <Docs>
-        <param name="bindingAttr">A bitmask comprised of one or more <see cref="T:System.Reflection.BindingFlags" /> that specify how the search is conducted.  
+        <param name="bindingAttr">A bitmask comprised of one or more enumeration members that specify how the search is conducted.
```
Comment (mairaw) [inline xml/System/Type.xml]:
```suggestion
        <param name="bindingAttr">A bitwise combination of the enumeration values that specify how the search is conducted.  
```

File: xml/System/Type.xml [hunk 2]
```diff
@@ -2950,11 +2949,11 @@ TopNamespace.Sub\+Namespace.ContainingClass+NestedClass, MyAssembly, Version=1.3
       </Parameters>
       <Docs>
         <param name="name">The string containing the name of an event which is declared or inherited by the current <see cref="T:System.Type" />.</param>
-        <param name="bindingAttr">A bitmask comprised of one or more <see cref="T:System.Reflection.BindingFlags" /> that specify how the search is conducted.  
+        <param name="bindingAttr">A bitmask comprised of one or more enumeration members that specify how the search is conducted.
```
Comment (mairaw) [inline xml/System/Type.xml]:
```suggestion
        <param name="bindingAttr">A bitwise combination of the enumeration values that specify how the search is conducted.  
```

File: xml/System/Type.xml [hunk 3]
```diff
@@ -3137,11 +3136,11 @@ TopNamespace.Sub\+Namespace.ContainingClass+NestedClass, MyAssembly, Version=1.3
         <Parameter Name="bindingAttr" Type="System.Reflection.BindingFlags" Index="0" FrameworkAlternate="netcore-2.0;netcore-2.1;netcore-2.2;netcore-3.0;netframework-1.1;netframework-2.0;netframework-3.0;netframework-3.5;netframework-4.0;netframework-4.5;netframework-4.5.1;netframework-4.5.2;netframework-4.6;netframework-4.6.1;netframework-4.6.2;netframework-4.7;netframework-4.7.1;netframework-4.7.2;netframework-4.8;netstandard-2.0;xamarinandroid-7.1;xamarinios-10.8;xamarinmac-3.0;netstandard-2.1" />
       </Parameters>
       <Docs>
-        <param name="bindingAttr">A bitmask comprised of one or more <see cref="T:System.Reflection.BindingFlags" /> that specify how the search is conducted.  
-  
- -or-  
-  
- Zero, to return <see langword="null" />.</param>
+        <param name="bindingAttr">A bitmask comprised of one or more enumeration members that specify how the search is conducted.
```
Comment (mairaw) [inline xml/System/Type.xml]:
```suggestion
        <param name="bindingAttr">A bitwise combination of the enumeration values that specify how the search is conducted.
```

File: xml/System/Type.xml [hunk 4]
```diff
@@ -3311,11 +3310,11 @@ TopNamespace.Sub\+Namespace.ContainingClass+NestedClass, MyAssembly, Version=1.3
       </Parameters>
       <Docs>
         <param name="name">The string containing the name of the data field to get.</param>
-        <param name="bindingAttr">A bitmask comprised of one or more <see cref="T:System.Reflection.BindingFlags" /> that specify how the search is conducted.  
+        <param name="bindingAttr">A bitmask comprised of one or more enumeration members that specify how the search is conducted.
```
Comment (mairaw) [inline xml/System/Type.xml]:
```suggestion
        <param name="bindingAttr">A bitwise combination of the enumeration values that specify how the search is conducted.  
```

File: xml/System/Type.xml [hunk 5]
```diff
@@ -1405,14 +1406,18 @@ TopNamespace.Sub\+Namespace.ContainingClass+NestedClass, MyAssembly, Version=1.3
...
   
+If the current <xref:System.Type> represents a type parameter of a generic type or generic method, <xref:System.Type.FindMembers%2A> processes any members declared by the class constraint and the interface constraints of the type parameter.  
+
+The `filter` argument can be a custom delegate of type <xref:System.Reflection.MemberFilter>, or it can be one of the following  predefined delegates:
+
+- <xref:System.Type.FilterAttribute?displayProperty=nameWithType>, which uses a<xref:System.Reflection.FieldAttributes>, <xref:System.Reflection.MethodAttributes>, or <xref:System.Reflection.MethodImplAttributes> bitmask as the `filterCriteria` value.
+
+- <xref:System.Type.FilterName?displayProperty=nameWithType>, which performs a case-sensitive comparison of each member name with the string passed to `filterCriteria`.
+
+- <xref:System.Type.FilterNameIgnoreCase?displayProperty=nameWithType>, which performs a case-insensitive comparison of each member name with the string passed to `filterCriteria`.
```
[the remark this answers is not in the document]
Comment (rpetrusha) [reply xml/System/Type.xml]: It seems to be working now.

**Comment by rpetrusha** [comment]
I wasn't able to batch merge your suggestions, @mairaw, so I did them in a separate local commit. Thanks for the suggestions.

**Pull request merged**